### ISSUE TITLE
PathWatcher does not work with relative files

### ISSUE THREAD

**Issue by scf37** [swoval/swoval#106]
When registering files with relative paths, change notification is expected to work and change event contain the same path passed in.

```scala
    val pw: PathWatcher[PathWatchers.Event] = PathWatchers.get(true)

    pw.addObserver(new Observer[PathWatchers.Event] {
      override def onError(t: Throwable): Unit = ???

      override def onNext(t: PathWatchers.Event): Unit = println(t)
    })
    val p = Paths.get("1")
    pw.register(p, 1)
    Files.write(p, "hello".getBytes())
    Thread.sleep(1000)
```

**Comment by eatkins** [comment]
Hmm...this is an interesting edge case. I have a fix to support it, but I need to think about what kinds of tests I need to write to ensure no unintended consequences. In the meantime, I do find that things work as expected if you replace `val p = Paths.get("1")` with `val p = Paths.get("1").toAbsolutePath` but I'm guessing you may have tried that already.

**Comment by scf37** [comment]
Yep. Idea is to keep paths as-is so library user can easily match registered paths and path available in event.

**Comment by eatkins** [comment]
I see. While writing this library, I was only interested in getting the full absolute path in the events and I never cared where the events came from. I think your use case is a new feature that I could add for a 2.1.0 release. I would likely add a new event class with the following api (defined in scala though the canonical implementation is in java):
```scala
trait RelativeEvent {
  def getRegistrationPath: Path
  def getKind: Kind
 // This would probably be provided as an absolute path, but getRegistrationPath would be the parent of this path
  def getTypedPath: TypedPath
}
```
which differs from the existing event class:
```scala
trait Event {
  def getKind: Kind
  def getTypedPath: TypedPath
}
```
Then I'd add a method `getRelative` or something to PathWatchers that returns `PathWatcher[RelativeEvent]` instead of `PathWatcher[Event]`.
Do you think that would meet your use case?

**Comment by scf37** [comment]
I'd prefer single API both for relative and absolute paths.

Why not just ensure that `TypedPath.path`  always starts with registration path? That keeps current behavior for absolute paths and makes the library more predictable for the user. If user wants absolute path or normalized path, he always can call appropriate method on `TypedPath.path`. Of course, if native implementations allows that for all platforms.

**Comment by eatkins** [comment]
The api would still be the same. It would just be the factory method for creating a PathWatcher that would be different. If you don't care at all about relative vs. absolute paths, you could continue using PathWatchers.get, otherwise you could make a new watcher with PathWatchers.getRelative and both absolute and relative paths would work fine; you would just get additional information in the event callback if the watcher was implemented with PathWatchers.getRelative.

From an implementation perspective, there isn't necessarily a 1:1 mapping of events to registration path. For example, it is possible to register foo with depth 1 and foo/bar with depth 0. If foo/bar/baz.txt is modified, what should the registration path for the event be?

My philosophy when writing this library was to keep it as minimal as possible (which is part of why it is implemented in java) while providing all of the needed functionality. There are a lot of subtleties with relative vs. absolute paths that I think are generally handled best at the application level or in a supporting library that provides a simpler interface. For example, I wrote a gist: [RelativePathWAtcher](https://gist.github.com/eatkins/b8f7e7ece77cba120830c1ce65b9e240) that has a crude implementation of a PathWatcher with similar semantics to what you described:

```scala
scala> import com.swoval.files._; import java.nio.file._; import FileTreeViews._
import com.swoval.files._
import java.nio.file._
import FileTreeViews._

scala> val pw = new RelativePathWatcher                       
pw: com.swoval.files.RelativePathWatcher = com.swoval.files.RelativePathWatcher@cf0669c

scala> pw.addObserver(new Observer[(Path, Path)] { override def onError(t: Throwable): Unit = Unit; override def onNext(e: (Path, Path)) = println(e) })
res28: Int = 0

scala> val file = Paths.get("1")                              
file: java.nio.file.Path = 1

scala> pw.register(file, 1)
res29: com.swoval.functional.Either[java.io.IOException,pw.JBool] = Right(true)

scala> Files.write(file, "foo".getBytes)                      
res30: java.nio.file.Path = 1

scala> val pa(1,)
rent = file.toAbsolutePath.getParent                          
parent: java.nio.file.Path = /Users/ethanatkins/work/swoval/files

scala> pw.register(parent, 1)
res31: com.swoval.functional.Either[java.io.IOException,pw.JBool] = Right(true)

scala> val parentFile = parent.resolve("bar")                 
parentFile: java.nio.file.Path = /Users/ethanatkins/work/swoval/files/bar

scala> Files.write(parentFile, "bar".getBytes)                
res32: java.nio.file.Path = /Users/ethanatkins/work/swoval/files/bar

scala> (/Users/ethanatkins/work/swoval/files,bar)  
```

**Comment by eatkins** [comment]
I want to add that I really appreciate your feedback. I somewhat aggressively suggested that you try my library and I don't want to be too defensive about the ways in which it doesn't exactly match your use case. The question is just how to reconcile the competing needs in a way that makes sense. I have been planning to eventually write a scala library on top of the core and I'd be more excited about adding alternative PathWatcher implementations (or even changing some of the core semantics) in the scala library instead of in the core java library.

**Comment by scf37** [comment]
I see that `NioPathWatcher` passes path to `WatchService` exactly how it was given to `register`.
May be if you fix bugs relating to improper handling of incorrect paths (like path == null in assert at NioPathWatcher:196) it will just work?

**Comment by eatkins** [comment]
Yeah. I guess what I've been trying to articulate is that I don't want to change any of the the internal data structures to use relative paths. I have no problem with converting relative paths to absolute paths at api boundaries, however. I'll try to get a release out next week that handles relative paths correctly.

**Comment by eatkins** [comment]
(actually there are many places that relative paths are used internally but there is an implicit assumption that users always pass in absolute paths at the api boundaries which is a needless restriction)

**Comment by eatkins** [comment]
I released version 2.0.8, which should work with relative paths (though the callbacks will still contain absolute paths).

**Comment by eatkins** [comment]
It may take a little while for the release to sync to maven.
